File: samples/jpeg_encode/jpeg_encode.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

import argparse
import vaststreamx as vsx


def argument_parser():
    parser = argparse.ArgumentParser("JPEG_ENCODER")
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--height",
        type=int,
        default=354,
        help="image height",
    )
    parser.add_argument(
        "--width",
        type=int,
        default=474,
        help="image width",
    )
    parser.add_argument(
        "--input_file",
        default="../../../data/images/cat_354x474_nv12.yuv",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="jpeg_encode_result.jpg",
        help="output file",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    # init card
    vsx.set_device(args.device_id)

    encoder = vsx.JpegEncoder()

    # prepare input data
    vsxImage = vsx.create_image(
        args.input_file,
        vsx.ImageFormat.YUV_NV12,
        args.width,
        args.height,
        vsx.Context.CPU(),
    )
    encoder.send_image(vsxImage)
    encoder.stop_send_image()

    # get output bytes
    output = encoder.recv_data()
    print(f"Encoded data bytes: {len(output)}")

    # write encoded bytes to output file
    with open(args.output_file, "wb") as f:
        f.write(output)

File: samples/jpeg_encode/jpeg_encode.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <fstream>
#include <iostream>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<uint32_t>("height", '\0', "image height", false, 354);
  args.add<uint32_t>("width", '\0', "image width ", false, 474);
  args.add<std::string>("input_file", '\0', "input image file", false,
                        "../data/images/cat_354x474_nv12.yuv");
  args.add<std::string>("output_file", '\0', "output image file", false,
                        "./jpeg_encode_result.jpg");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  int width = args.get<uint32_t>("width");
  int height = args.get<uint32_t>("height");
  // init card
  vsx::SetDevice(args.get<uint32_t>("device_id"));

  vsx::JpegEncoder encoder;

  // prepare data
  std::shared_ptr<vsx::DataManager> data;
  vsx::ReadBinaryFile(args.get<std::string>("input_file"), data);

  // create cpu vsx::Image, width_pith,height_pitch use default,
  vsx::Image vsx_image(vsx::ImageFormat::YUV_NV12, width, height, 0, 0, data);

  // start encode
  encoder.SendImage(vsx_image);
  encoder.StopSendImage();

  // get encoded bytes
  std::shared_ptr<vsx::DataManager> encoded_bytes;
  encoder.RecvData(encoded_bytes);
  std::cout << "Encoded data bytes: " << encoded_bytes->GetDataSize() << "\n";

  // write encoded bytes to output file
  vsx::WriteBinaryFile(args.get<std::string>("output_file"),
                       encoded_bytes->GetDataPtr(),
                       encoded_bytes->GetDataSize());

  return 0;
}